### PR TITLE
Completed impl of double tap

### PR DESCRIPTION
**Pull request by ocean-ift**
Unit tests include tests for up to three finger double taps

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (xorbit3024) [inline .vscode/settings.json:1]: Please .gitignore the .vscode folder.

File: gesturelibrary/include/double.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+/// @brief
+/// @param event
+gesture_event_t* recognize_double_tap(touch_event_t* event);
+
+/// @brief
+/// @param _double
+/// @param strokes
+/// @param taps
+void update_double_taps(double_tap_t* _double, stroke_t* strokes, tap_t* taps);
```
Comment (xorbit3024) [inline gesturelibrary/include/double.h:30]: If it doesn't need to be exposed to the user, please declare it static in the .c file instead.

File: gesturelibrary/include/double.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+
+/// @brief
+/// @param c1
+/// @param c2
+/// @return
+int euclidean_distance(coords_t c1, coords_t c2);
```
Comment (xorbit3024) [inline gesturelibrary/include/double.h:36]: If it doesn't need to be exposed to the user, please declare it static in the .c file instead.

File: gesturelibrary/include/double.h [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    int y;
+    int t;
+} double_tap_t;  // single finger double tap data
+
+typedef struct coords {
+    int x;
+    int y;
+} coords_t;
+
+/// @brief
```
Comment (xorbit3024) [inline gesturelibrary/include/double.h:22]: Please fill out doxygen comments.

File: gesturelibrary/include/double.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+/// @param event
+gesture_event_t* recognize_double_tap(touch_event_t* event);
+
+/// @brief
+/// @param _double
+/// @param strokes
+/// @param taps
+void update_double_taps(double_tap_t* _double, stroke_t* strokes, tap_t* taps);
+
+/// @brief
```
Comment (xorbit3024) [inline gesturelibrary/include/double.h:32]: Please fill out doxygen comments.

File: gesturelibrary/include/double.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
+#pragma once
+
+#include "gesturelib.h"
+#include "math.h"
+#include "recognizer.h"
+#include "stroke.h"
+#include "tap.h"
+
+/// @brief
```
Comment (xorbit3024) [inline gesturelibrary/include/double.h:9]: Please fill out doxygen comments.

File: gesturelibrary/include/double.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+typedef struct double_tap {
+    state_t state;
+    int x;
+    int y;
+    int t;
+} double_tap_t;  // single finger double tap data
+
+typedef struct coords {
```
Comment (xorbit3024) [inline gesturelibrary/include/double.h:17]: If it doesn't need to be exposed to the user, please declare it in the .c file instead.

File: gesturelibrary/src/double.c [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+        // if too far, move on to the next one
+        // if timestamp is too far, set this double to failed
+        if (strokes->state == RECOGNIZER_STATE_COMPLETED) {
+            coords_t c1 = {.x = _double->x, .y = _double->y};
+            coords_t c2 = {.x = strokes->y, .y = strokes->y};
+            int dist    = euclidean_distance(c1, c2);
+
+            if (_double->t - strokes->t >
+                DOUBLE_DIFF) {  // if t_diff too big, this dTap has failed cuz we're not getting another stroke that
+                                // magically went back in time, ya feel?
```
Comment (xorbit3024) [inline gesturelibrary/src/double.c:38]: Please make professional comments.

File: gesturelibrary/test/res/hold/phone_1.csv [hunk 9]
```diff
@@ -1,3 +1,3 @@
-type,down,position.dx,position.dy,timeStamp
-down,true,220,243,0:00:03.178000
+type,down,position.dx,position.dy,timeStamp
+down,true,220,243,0:00:03.178000
 up,false,220,243,0:00:04.660000
```
Comment (xorbit3024) [inline gesturelibrary/test/res/hold/phone_1.csv:3]: Please ensure all files end with \n.

File: gesturelibrary/test/res/hold/phone_2.csv [hunk 10]
```diff
@@ -1,7 +1,7 @@
...
-move,true,225,232,0:00:04.819000
-move,true,226,232,0:00:06.385000
-move,true,227,233,0:00:07.134000
+type,down,position.dx,position.dy,timeStamp
+down,true,231,231,0:00:01.869000
+move,true,224,232,0:00:03.236000
+move,true,225,232,0:00:04.819000
+move,true,226,232,0:00:06.385000
+move,true,227,233,0:00:07.134000
 up,false,227,233,0:00:07.134000
```
Comment (xorbit3024) [inline gesturelibrary/test/res/hold/phone_2.csv:7]: Please ensure all files end with \n.

File: gesturelibrary/test/res/hold/phone_3.csv [hunk 11]
```diff
@@ -1,3 +1,3 @@
-type,down,position.dx,position.dy,timeStamp
-down,true,212,294,0:00:01.548000
+type,down,position.dx,position.dy,timeStamp
+down,true,212,294,0:00:01.548000
 up,false,212,294,0:00:02.606000
```
Comment (xorbit3024) [inline gesturelibrary/test/res/hold/phone_3.csv:3]: Please ensure all files end with \n.

File: gesturelibrary/test/res/hold/phone_4.csv [hunk 12]
```diff
@@ -1,7 +1,7 @@
...
-move,true,231,242,0:00:02.984000
-move,true,232,242,0:00:03.276000
-move,true,232,240,0:00:04.266000
+type,down,position.dx,position.dy,timeStamp
+down,true,231,243,0:00:02.284000
+move,true,230,242,0:00:02.917000
+move,true,231,242,0:00:02.984000
+move,true,232,242,0:00:03.276000
+move,true,232,240,0:00:04.266000
 up,false,232,240,0:00:04.266000
```
Comment (xorbit3024) [inline gesturelibrary/test/res/hold/phone_4.csv:7]: Please ensure all files end with \n.

File: gesturelibrary/test/res/hold/phone_5.csv [hunk 13]
```diff
@@ -1,5 +1,5 @@
-type,down,position.dx,position.dy,timeStamp
-down,true,227,278,0:00:01.273000
-move,true,229,277,0:00:02.290000
-move,true,230,277,0:00:02.682000
+type,down,position.dx,position.dy,timeStamp
+down,true,227,278,0:00:01.273000
+move,true,229,277,0:00:02.290000
+move,true,230,277,0:00:02.682000
 up,false,230,277,0:00:02.682000
```
Comment (xorbit3024) [inline gesturelibrary/test/res/hold/phone_5.csv:5]: Please ensure all files end with \n.

File: gesturelibrary/test/res/hold/phone_7.csv [hunk 14]
```diff
@@ -1,10 +1,10 @@
...
+type,down,position.dx,position.dy,timeStamp
+down,true,282,210,0:00:03.054000
+down,true,181,294,0:00:03.079000
+move,true,179,305,0:00:03.596000
+move,true,178,305,0:00:03.671000
+move,true,277,213,0:00:03.671000
+move,true,276,214,0:00:03.671000
+up,false,276,214,0:00:03.671000
+move,true,178,302,0:00:03.753000
 up,false,178,302,0:00:03.753000
```
Comment (xorbit3024) [inline gesturelibrary/test/res/hold/phone_7.csv:10]: Please ensure all files end with \n.

**Comment by xorbit3024** [comment]
Consider the following situation. 1 finger performs a tap, then 2nd finger comes down. Without letting go of 2nd finger, 1st finger performs tap again. Suppose this all happens fast enough for the 1st finger's two taps to make sense as a double tap. Does your implementation account for this? If so, just let me know, maybe I can't tell from the code. If not, please consider adding this consideration or posting it as a known issue for another story.

File: .vscode/settings.json (hunk 1 above)
Comment (Russell-Newton) [reply .vscode/settings.json:1]: you'll need to delete this folder and push that deletion to git before the .gitignore will work for it

**Pull request closed without merging**